client/starwhale/core/job/scheduler.py

```diff
@@ -0,0 +1,258 @@
+                    _finished_step_names.append(_step.step_name)
+                if _step.status is STATUS.INIT:
+                    _wait_steps.append(_step)
+            # judge whether a step's dependency all in finished
+            for _wait in _wait_steps:
+                if all(d in _finished_step_names for d in _wait.dependency if d):
+                    _wait.status = STATUS.RUNNING
+                    # add pipe handler
+
+                    for _t in _wait.tasks:
+                        _d = TaskDaemon(
+                            TaskPipe(
+                                idx=f"{_wait.step_name}-{_t.context.index}",
+                                main_conn=_t.main_conn,
+                                sub_conn=_t.sub_conn,
+                            ),
+                            self,
+                        )
+                        _d.start()
+                    _executor = Executor(
+                        _wait.concurrency, _wait, _wait.tasks, StepCallback(self)
+                    )
+                    _executor.start()
+                    # executor.setDaemon()
+                    _threads.append(_executor)
+
+        for _thread in _threads:
+            _thread.join()
+
+    def schedule_single_task(self, step_name: str, task_index: int) -> None:
+        _steps = [step for step in self.steps if step_name == step.step_name]
+        if len(_steps) == 0:
+            raise RuntimeError(f"step:{step_name} not found")
+        _step = _steps[0]
+        if task_index >= _step.task_num:
+            raise RuntimeError(
+                f"task_index:{task_index} out of bounds, total:{_step.task_num}"
+            )
+        _task = _step.tasks[task_index]
+        logger.debug(
+            "all result size: {}", len([res for res in self._datastore.get_results()])
+        )
+        _d = TaskDaemon(
+            TaskPipe(
+                idx=f"{_step.step_name}-{_task.context.index}",
+                main_conn=_task.main_conn,
+                sub_conn=_task.sub_conn,
+            ),
+            self,
+        )
+        _d.start()
+
+        _executor = Executor(1, _step, [_task], SingleTaskCallback(self))
+        _executor.start()
+        _executor.join()
+
+
+class TaskDaemon(threading.Thread):
+    def __init__(self, task_pipe: TaskPipe, scheduler: Scheduler):
+        super().__init__()
+        self.scheduler = scheduler
+        self.task_pipe = task_pipe
+        self.main_conn = task_pipe.main_conn
+        self.setDaemon(True)
+
+    def run(self) -> None:
+        while True:
+            logger.debug("task:{} start to waiting data", self.task_pipe.id)
+            data = self.main_conn.recv()
+            logger.debug("task:{} start to recv data", self.task_pipe.id)
+            if isinstance(data, EvaluationQuery):
+                self.main_conn.send(self.scheduler.query_data(data))
+            else:
+                self.scheduler.add_data(data)
+
+
+class Callback:
```

Comment:
Using the duck typing is the better choice. 

https://adamj.eu/tech/2021/05/18/python-type-hints-duck-typing-with-protocol/
example code:  https://github.com/star-whale/starwhale/blob/main/client/starwhale/api/_impl/data_store.py#L728

Reply:
done